File: common/src/main/java/juuxel/adorn/client/gui/screen/PalettedMenuScreen.java
package juuxel.adorn.client.gui.screen;

import juuxel.adorn.client.resources.ColorManager;
import juuxel.adorn.menu.ContainerBlockMenu;
import net.minecraft.client.gui.GuiGraphicsExtractor;
import net.minecraft.client.renderer.RenderPipelines;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.Identifier;
import net.minecraft.world.entity.player.Inventory;
import net.minecraft.world.inventory.AbstractContainerMenu;
import net.minecraft.world.level.block.Blocks;

public abstract class PalettedMenuScreen<M extends AbstractContainerMenu & ContainerBlockMenu> extends AdornMenuScreen<M> {
    private final Identifier blockId;

    public PalettedMenuScreen(M menu, Inventory playerInventory, Component title) {
        super(menu, playerInventory, title);
        this.blockId = BuiltInRegistries.BLOCK.getKey(menu.getContext().evaluate((world, pos) -> world.getBlockState(pos).getBlock(), Blocks.AIR));
    }

    protected abstract Identifier getBackgroundTexture();
    protected abstract Identifier getPaletteId();

    private ColorManager.ColorPair getPalette() {
        return ColorManager.INSTANCE.getColors(getPaletteId()).get(blockId);
    }

    @Override
    public void extractBackground(GuiGraphicsExtractor context, int mouseX, int mouseY, float a) {
        super.extractBackground(context, mouseX, mouseY, a);
        var bg = getPalette().bg();
        context.blit(RenderPipelines.GUI_TEXTURED, getBackgroundTexture(), leftPos, topPos, 0, 0, imageWidth, imageHeight, 256, 256, bg);
    }

    @Override
    protected void extractLabels(GuiGraphicsExtractor context, int mouseX, int mouseY) {
        var fg = getPalette().fg();
        context.text(font, title, titleLabelX, titleLabelY, fg, false);
        context.text(font, playerInventoryTitle, inventoryLabelX, inventoryLabelY, fg, false);
    }
}

File: common/src/main/kotlin/juuxel/adorn/client/gui/screen/PalettedMenuScreen.kt
package juuxel.adorn.client.gui.screen

import com.mojang.blaze3d.systems.RenderSystem
import juuxel.adorn.menu.ContainerBlockMenu
import juuxel.adorn.platform.PlatformBridges
import juuxel.adorn.util.Colors
import juuxel.adorn.util.getBlock
import net.minecraft.client.gui.DrawContext
import net.minecraft.entity.player.PlayerInventory
import net.minecraft.menu.Menu
import net.minecraft.registry.Registries
import net.minecraft.text.Text
import net.minecraft.util.Identifier

abstract class PalettedMenuScreen<M>(menu: M, playerInventory: PlayerInventory, title: Text) :
    AdornMenuScreen<M>(menu, playerInventory, title)
    where M : Menu, M : ContainerBlockMenu {
    protected abstract val backgroundTexture: Identifier
    protected abstract val paletteId: Identifier
    private val blockId = Registries.BLOCK.getId(menu.context.getBlock())

    private fun getPalette() = PlatformBridges.resources.colorManager.getColors(paletteId)[blockId]

    override fun drawBackground(context: DrawContext, delta: Float, mouseX: Int, mouseY: Int) {
        val bg = getPalette().bg
        RenderSystem.setShaderColor(Colors.redOf(bg), Colors.greenOf(bg), Colors.blueOf(bg), 1.0f)
        context.drawTexture(backgroundTexture, x, y, 0, 0, backgroundWidth, backgroundHeight)
        RenderSystem.setShaderColor(1f, 1f, 1f, 1f)
    }

    override fun drawForeground(context: DrawContext, mouseX: Int, mouseY: Int) {
        val fg = getPalette().fg
        context.drawText(textRenderer, title, titleX, titleY, fg, false)
        context.drawText(textRenderer, playerInventoryTitle, playerInventoryTitleX, playerInventoryTitleY, fg, false)
    }
}
